### PR TITLE
Number struct in Expressions

### PR DESCRIPTION
**Pull request by gzanitti**
This PR is part of issue https://github.com/powdr-labs/powdr/pull/1345.
In particular, it adds the struct Number to Expressions to homogenize the structure before including source references.

### CODE REVIEW

File: ast/src/parsed/mod.rs [hunk 1]
```diff
@@ -385,13 +398,17 @@ impl Expression<NamespacedPolynomialReference> {
 
 impl From<u32> for Expression {
     fn from(value: u32) -> Self {
-        Expression::Number(value.into(), None)
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:388]: Can you call From<BigUint> here?

File: executor/src/constant_evaluator/mod.rs [hunk 2]
```diff
@@ -71,7 +71,13 @@ fn generate_values<T: FieldElement>(
             let e = if let Some(index) = index {
                 index_expr = Expression::IndexAccess(IndexAccess {
                     array: e.clone().into(),
-                    index: Box::new(Expression::Number(index.into(), None)),
+                    index: Box::new(
```
Comment (chriseth) [inline executor/src/constant_evaluator/mod.rs:74]: Can you directly use index.into() here?

File: linker/src/lib.rs [hunk 3]
```diff
@@ -141,12 +141,12 @@ fn process_link(link: Link) -> PilStatement {
...
 
     // the lhs is `instr_flag { operation_id, inputs, outputs }`
-    let op_id = to
-        .operation
-        .id
-        .iter()
-        .cloned()
-        .map(|n| Expression::Number(n, None));
+    let op_id = to.operation.id.iter().cloned().map(|n| {
+        Expression::Number(Number {
```
Comment (chriseth) [inline linker/src/lib.rs:145]: n.into()?

File: parser/src/powdr.lalrpop [hunk 4]
```diff
@@ -578,7 +578,7 @@ Term: Box<Expression> = {
     ConstantIdentifier => Box::new(Expression::Reference(NamespacedPolynomialReference::from_identifier(<>))),
     GenericReference => Box::new(Expression::Reference(<>)),
     PublicIdentifier => Box::new(Expression::PublicReference(<>)),
-    Number => Box::new(Expression::Number(<>.into(), None)),
+    Number => Box::new(Number{value: <>.into(), type_:None}.into()),
```
Comment (chriseth) [inline parser/src/powdr.lalrpop:581]:
```suggestion
    Number => Box::new(Number{value: <>.into(), type_: None}.into()),
```

File: pil-analyzer/src/expression_processor.rs [hunk 5]
```diff
@@ -87,7 +87,9 @@ impl<'a, D: AnalysisDriver> ExpressionProcessor<'a, D> {
         match expr {
             PExpression::Reference(poly) => Expression::Reference(self.process_reference(poly)),
             PExpression::PublicReference(name) => Expression::PublicReference(name),
-            PExpression::Number(n, t) => Expression::Number(n, t),
+            PExpression::Number(Number { value: n, type_: t }) => {
+                Number { value: n, type_: t }.into()
+            }
```
Comment (chriseth) [inline pil-analyzer/src/expression_processor.rs:92]:
```suggestion
            PExpression::Number(number) => number.into(),
```

File: asm-to-pil/src/romgen.rs [hunk 6]
```diff
@@ -49,7 +49,13 @@ fn pad_return_arguments(s: &mut FunctionStatement, output_count: usize) {
     if let FunctionStatement::Return(ret) = s {
         ret.values = std::mem::take(&mut ret.values)
             .into_iter()
-            .chain(repeat(Expression::Number(0u32.into(), None)))
+            .chain(repeat(
+                Number {
+                    value: 0u32.into(),
```
Comment (chriseth) [inline asm-to-pil/src/romgen.rs:54]: 0u32.into() directly?

Comment (chriseth) [reply asm-to-pil/src/romgen.rs:54]: I think there are several more places, won't comment on all of them.

File: asm-to-pil/src/vm_to_constrained.rs [hunk 7]
```diff
@@ -684,13 +688,13 @@ impl<T: FieldElement> VMConverter<T> {
                         }
                         Input::Literal(_, LiteralKind::SignedConstant) => {
                             // TODO evaluate expression
-                            if let Expression::Number(n, _) = a {
+                            if let Expression::Number(Number {value: n, type_: _}) = a {
```
Comment (chriseth) [inline asm-to-pil/src/vm_to_constrained.rs:691]: Formatting? (also below)

File: ast/src/parsed/mod.rs [hunk 8]
```diff
@@ -442,7 +454,7 @@ impl<R> Expression<R> {
             Expression::Reference(_) | Expression::PublicReference(_) | Expression::String(_) => {
                 empty()
             }
-            Expression::Number(_, _) => empty(),
+            Expression::Number(_) => empty(),
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:457]: I'm wondering, should we implement `Children<Expression>` on all the structs and then uniformly call `.children()` on all of them here? Essentially saying that `Expression` only knows the collection of enums it has, but doesn't care what the inner structs look like and what they do. We could even encapsulate that in a macro "looping" over all the enum values.

Comment (gzanitti) [reply ast/src/parsed/mod.rs:457]: I just answered you something similar on MatchExpression's PR.  I agree with you, the implementation of children for expression should simply use the children of each variant.

**Review by chriseth — APPROVED** [review]

**Pull request merged**